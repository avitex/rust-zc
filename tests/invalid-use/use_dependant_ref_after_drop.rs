use zc::Dependant;

#[derive(Dependant, PartialEq, Debug)]
pub struct MyStruct<'a>(&'a [u8]);

impl<'a> From<&'a [u8]> for MyStruct<'a> {
    fn from(bytes: &'a [u8]) -> Self {
        Self(&bytes)
    }
}

fn main() {
    let owner = vec![1, 2, 3];
    let data = zc::from!(owner, MyStruct, [u8]);
    let dependant_ref = data.dependant::<MyStruct>();
    core::mem::drop(data);
    assert_eq!(
        dependant_ref,
        &MyStruct(&[1, 2, 3])
    )
}
