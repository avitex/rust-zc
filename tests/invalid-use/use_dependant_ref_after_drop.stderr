error[E0505]: cannot move out of `data` because it is borrowed
  --> $DIR/use_dependant_ref_after_drop.rs:16:21
   |
15 |       let dependant_ref = data.dependant::<MyStruct>();
   |                           ---- borrow of `data` occurs here
16 |       core::mem::drop(data);
   |                       ^^^^ move out of `data` occurs here
17 | /     assert_eq!(
18 | |         dependant_ref,
19 | |         &MyStruct(&[1, 2, 3])
20 | |     )
   | |_____- borrow later used here
